Open list of all types

Woyten opened the pull request:
As announced, I worked on a feature similar to the _Open Type_ feature for Java. This is my first time working with Eclipse PDE and MelnormeLang. So don't be mad at me if I did not use the whole infrastructure as intended. ;-)

The feature is not yet completely finished but as it is in a usable state, I would like to hear your opinion.
Things still missing are:
- Non-blocking calculations in the background 
- Unit tests
- Cosmetic issues (e.g. camel case search) that could also be part of a separate issue

During coding, I used Eclipse auto-formatting and some save actions. The reason is that I am strongly used to auto-formatting and my code would just not look nice without it. I will provide you the auto-format settings in another pull request.

Woyten commented on plugin_ide.core.tests/src-lang/melnorme/lang/ide/core/engine/StructureModelTest.java:
```diff
@@ -60,8 +60,7 @@ public FixtureSourceModelManager() {
 		
-		public final SourceFileStructure DEFAULT_STRUCTURE = new SourceFileStructure(null, null, 
-			(Indexable<ParserError>) null);
+		public final SourceFileStructure DEFAULT_STRUCTURE = new SourceFileStructure(null, (Indexable<ParserError>) null);
```
I removed the first parameter of `SourceFileStructure` as the code location is now stored in `StructureElement`.

bruno-medeiros replied:
Why was this done in the first place? It seems unnecessary.

bruno-medeiros replied:
Yeah this is definitely wrong. You are adding a location field to every StructureElement in a tree, even though the location is the same for all in the tree. I've added: https://github.com/RustDT/RustDT/commit/fe73508dce4baba8844e7ac2762440c3fba18ffd

Woyten replied:
True. In one source file, the location is always the same. However, for the index, the `StructureElement`s can be different. There are several solutions for that problem. The one I committed seemed to be the most straight-forward to me. What do you think?

bruno-medeiros replied:
What do I think? I'm not sure what you mean. Yes there can be StructureElements with different locations, but I don't see what is the issue. Just use the getLocation method as added in https://github.com/RustDT/RustDT/commit/fe73508dce4baba8844e7ac2762440c3fba18ffd Note that for that method to work, you can't clone the StructureElements as you were doing before, but I see you already changed that so that they are not cloned, as suggested.

Woyten replied:
You're right. There is no issue. My bad!
I merged your changes and restored the `StructureElement`s to their original shape.

Woyten commented on plugin_ide.core/src/com/github/rustdt/ide/core/engine/RustSourceModelManager.java:
```diff
@@ -57,49 +52,32 @@ protected SourceFileStructure doCreateNewData() throws CommonException, Operatio
+						return new SourceFileStructure(previousElements, newStructure.getParserProblems());
 					}
 				}
+				System.out.println("RustSourceModelManager - File touched: " + fileLocation);
```
My intention is to remove this line once we agree that the _Open Type_ feature works. A better solution would be to use debug logging. Do we have the infrastructure for that?

Woyten commented on plugin_ide.core/src/com/github/rustdt/ide/core/engine/RustSourceModelManager.java:
```diff
@@ -57,49 +52,32 @@ protected SourceFileStructure doCreateNewData() throws CommonException, Operatio
 					}
 				}
+				System.out.println("RustSourceModelManager - File touched: " + fileLocation);
+				GlobalSourceStructure.fileTouched(fileLocation, newStructure);
```
This is one location where the index for all types is filled.

Woyten commented on plugin_ide.core/src/com/github/rustdt/ide/core/engine/RustSourceModelManager.java:
```diff
@@ -57,49 +52,32 @@ protected SourceFileStructure doCreateNewData() throws CommonException, Operatio
-			
-		}
-		
-		protected String getDescribeOutput(String source, Location fileLocation) 
```
I had to reuse that part. So, I moved it to `RustParseDescribeLauncher`.

Woyten commented on plugin_ide.ui/plugin.xml:
```diff
@@ -179,6 +186,12 @@
 		</key>
 		<key
+			commandId="com.github.rustdt.ide.ui.commands.OpenType"
+			contextId="com.github.rustdt.ide.ui.Contexts.Editor"
```
This binding currently only works if an editor is open. Changing the `contextId` did not help. Any idea how to solve that problem?

Woyten commented on plugin_ide.ui/src-lang/melnorme/lang/ide/ui/editor/actions/OpenTypeHandler.java:
```diff
@@ -0,0 +1,50 @@
+import melnorme.lang.tooling.structure.GlobalSourceStructure;
+import melnorme.utilbox.core.CommonException;
+
+public class OpenTypeHandler extends AbstractEditorHandler {
```
I am not entirely familiar with the separation of `src` and `src-lang`. The package might be wrong.

Woyten commented on plugin_ide.ui/src-lang/melnorme/lang/ide/ui/editor/structure/EditorStructureUtil.java:
```diff
@@ -11,28 +11,29 @@
 package melnorme.lang.ide.ui.editor.structure;
 
 import static melnorme.lang.ide.ui.EditorSettings_Actual.EDITOR_ID;
-import melnorme.lang.ide.ui.editor.EditorUtils;
```
Eclipse rearranged those imports for me. I hope, you don't mind. From my point of view the change makes sense as `org` before `melnorme` seems to be the dominant package order in the code base. Please correct me if I'm wrong.

bruno-medeiros replied:
Yeah this is fine. We don't really look at imports, it's okay for Organize Imports action to reorder them. Most newer files are already organized that way, but that are some older files that haven't had the import headers organized yet.

Woyten commented on plugin_ide.ui/src/com/github/rustdt/ide/ui/editor/RustEditorActionContributor.java:
```diff
@@ -28,6 +47,88 @@ protected RustOpenDefinitionOperation createOpenDefinitionOperation(ITextEditor
 	
 	@Override
 	protected void registerOtherEditorHandlers() {
+		setupOpenType();
```
Not sure where to put that code. Here or in the `LangEditorActionContributor`? Or something else? The global type structure is not directly related to an editor, is it?

bruno-medeiros replied:
Yeah this is wrong. First, the code to manage the index should be in the core plugin, not UI. Second it needs to be initialized as one of the plugin singletons, on plugin startup. Like RustBundleModelManager for example.

As this will be closely related to SourceModelManager, it might be best to add the "IndexManager" as component of SourceModelManager.

Woyten replied:
Thanks for the hint! It helped me a lot understanding how the plugin is set up. As a result, I created the classes `RustIndexManager` and `IndexManager` as siblings to the existing `RustSourceModelManager` and `SourceModelManager`. But I did not include them to the `SourceModelManager` for two reasons:
1. Indexing and parsing/error checking are two different concepts that often behave completely differently and should remain separate in my eyes.
2. The structure of the class is a bit confusing. Especially those nested non-static classes drive me crazy ;-). Would be great if we could improve the readability here.
Besides that, I am wondering how the language agnostic _Open Type_ dialog will work in the end. We could just call `LangCore.getIndexManager()` but probably there is a better solution in place.

bruno-medeiros replied:
> Besides that, I am wondering how the language agnostic Open Type dialog will work in the end. We could just call LangCore.getIndexManager() but probably there is a better solution in place.

Yeah that is fine actually. If it comes that elsewhere we need to get `RustIndexManager` instead of just `IndexManager`, an accessor method can be created in LangCore_Actual that provides that. For example, that is done in DDT, for the ToolManager and BundleModelManager singletons, there are accessors with language-specific return types:
https://github.com/DDT-IDE/DDT/blob/1b68c437fd680d0478c7b97fe43c926d4da902b7/plugin_ide.core/src/melnorme/lang/ide/core/LangCore_Actual.java#L57
It may look a bit odd at times, but in practice it works fine.

Woyten commented on plugin_ide.ui/src/com/github/rustdt/ide/ui/editor/RustEditorActionContributor.java:
```diff
@@ -28,6 +47,88 @@ protected RustOpenDefinitionOperation createOpenDefinitionOperation(ITextEditor
+		setupOpenType();
+	}
+	
+	// TODO: Evaluate source structure in background thread.
```
I will be working on that. But if I could need some hints which infrastructure to reuse.

Woyten replied:
I'll try to reuse the `RustStructureUpdateTask`.

Woyten commented on plugin_ide.ui/src/com/github/rustdt/ide/ui/editor/RustEditorActionContributor.java:
```diff
@@ -28,6 +47,88 @@ protected RustOpenDefinitionOperation createOpenDefinitionOperation(ITextEditor
+	// TODO: Find a good filter
+	private static boolean isValidRustPath(IResource resource) {
+		String path = resource.getFullPath().toString();
+		return path.contains("src") && path.endsWith(".rs");
```
This way of filtering Rust source files is very lame. Is there some central logic that I can reuse? Or can I ask the Eclipse API to return only files that are part of the configured source folders?

Woyten replied:
Found out that `IResourceChangeListener`s are only notified about Rust file changes. Obviously, there is some integrated source file concept. Still, I don't know how to apply this concept for the initial indexing run.

bruno-medeiros replied:
Accepting any resource that ends with ".rs" should be fine for now, even if the file is not part of a source directory. This case should rare anyways, but even then, if it's a Rust file outside of a source folder one can say the user might still want to Open-Type into that file.

Woyten replied:
Yes. Scanning all rust files might be useful as this would also include external library files. However, ParseDescribe started blocking when large files of the glutin library were indexed. Maybe the situation gets better with better concurrency support.

Woyten replied:
Just tried it out without the source folder filtering. The performance is now reasonable.

Woyten commented on plugin_ide.ui/src/com/github/rustdt/ide/ui/editor/RustEditorActionContributor.java:
```diff
@@ -28,6 +47,88 @@ protected RustOpenDefinitionOperation createOpenDefinitionOperation(ITextEditor
+		return true;
+	}
+	
+	private void listenToUpdatesOfGlobalSourceStructure() {
```
Updates of the file system are dealt with here.

Woyten commented on plugin_tooling/src-lang/melnorme/lang/tooling/structure/SourceFileStructure_Default.java:
```diff
@@ -38,33 +34,22 @@ public boolean equals(Object obj) {
 	}
 	
 	@Override
 	public int hashCode() {
```
Is `equals` and `hashCode` really necessary here? If I am not mistaken, it is not. This is rather confusing as it gives the developer the impression that this part is important and he has to care about it. My preferred way, although a bit radical, is to throw `RuntimeException`s any time you call `equals` or `hashCode` s.t. you cannot use those methods in the wrong context.

Woyten commented on plugin_tooling/src/melnorme/lang/tooling/structure/GlobalSourceStructure.java:
```diff
@@ -0,0 +1,49 @@
+import melnorme.utilbox.misc.Location;
+
+public class GlobalSourceStructure {
+	private static final SortedMap<Location, SortedSet<StructureElement>> aggregatedElements = new TreeMap<>(comparing(Location::toPath));
```
I am not happy with having static instances here. Is there some alternative way, e.g. attaching instances to the workspace context?

Woyten commented on plugin_tooling/src/melnorme/lang/tooling/structure/GlobalSourceStructure.java:
```diff
@@ -0,0 +1,49 @@
+public class GlobalSourceStructure {
+	private static final SortedMap<Location, SortedSet<StructureElement>> aggregatedElements = new TreeMap<>(comparing(Location::toPath));
+	
+	private static final EnumSet<StructureElementKind> HIDDEN_ELEMENT_KINDS = EnumSet.of(StructureElementKind.EXTERN_CRATE,
```
Not sure if we should hide methods, impls, etc. as well. It's clearly something we should discuss about.

bruno-medeiros commented on plugin_ide.ui/src/com/github/rustdt/ide/ui/editor/RustEditorActionContributor.java:
```diff
@@ -28,6 +51,99 @@ protected RustOpenDefinitionOperation createOpenDefinitionOperation(ITextEditor
+				GlobalSourceStructure.fileTouched(location, fileStructure);
+			}
+		} catch(OperationCancellation | CommonException e) {
+			throw new RuntimeException(e);
```
No, no RuntimeExceptions (unless it's an assertion failure). See https://github.com/bruno-medeiros/MelnormeEclipse/wiki/Extensive-Compile-Time-Checking

Woyten replied:
Hmm... here, the problem with checked exceptions is that they are not well supported in lambdas and that they are also forbidden in the `Job.run()` method. As a compromise, we could just log here but that would be a duplication of line 69.

bruno-medeiros replied:
It's a problem that has to be handled somehow. One basic solution is indeed logging, for example using `LangCore.logError`. This is will show up in the Eclipse error log. It's a bit basic because it's hard to display that as an error somewhere else in the UI. 

A better solution would be to store the error in `GlobalSourceStructure` somehow, associated with it's Location. For example of something similar, the SourceModelManager stores for each location not just a `SourceFileStructure` but a  `CommonResult<SourceFileStructure>` which means it can store the error for that location, and so we are able to display the error in the Outline view. But I think all this can be added later, for now logging would be fine.

Woyten replied:
The problem has disappeared as I use the `DataUpdateTask` mechanism now.

bruno-medeiros commented on plugin_tooling/src-lang/melnorme/lang/tooling/structure/AbstractStructureContainer.java:
```diff
@@ -43,7 +46,23 @@ protected void _invariant() {
+		for(StructureElement child : getChildren()) {
+			flattenedElements.addAll(child.flattenTree());
+		}
+		return flattenedElements;
```
This should not be used as it forces the creation of an ArrayList, who is likely not to be used anyways. Better to have a Visitor method and visits/travels over the tree. This can then be used by another Collection to add all visited elements to it, without having to create an intermediate ArrayList.

Also, why are you cloning the StructureElement too? You can just add it directly to the collection, since StructureElement is an immutable object.

Woyten replied:
Sounds reasonable. Was not so happy about the flatten neither. I will think about it.

bruno-medeiros commented on plugin_ide.ui/src/com/github/rustdt/ide/ui/editor/RustEditorActionContributor.java:
```diff
@@ -28,6 +51,99 @@ protected RustOpenDefinitionOperation createOpenDefinitionOperation(ITextEditor
+			case IResourceDelta.REMOVED:
+				fileRemoved(location);
+				break;
+			}
```
For this to have correct concurrent behavior, you will need to analyze the delta in the same thread that triggers the delta. However, you cannot analyze the file itself in the delta thread, as that is too slow. What needs to be done is that GlobalSourceStructure (or a similar class) must take not immediately (in the delta thread) of which files have been modified/removed, and then queue a source structure update in some background thread. (Possibly cancelling the previous update if it is still running).  You can see an example of this in DocumentReconcileManager, which is responsible for listening to document changes (keypresses), and notifying the SourceModelManager that a new update is needed.

Also, this way the GlobalSourceStructure always knows which files are updated, and which are not (this can then be used in a UI display like the Open-Type dialog, where it can display a text like "Indexing 23 files...", in a similar way to what JDT does)

Woyten replied:
Are you sure about that? Do I get inconsistent data if I analyze the delta in a separate thread?
I will see what I can do anyway. What about iterating over the delta and scheduling one job per file?

bruno-medeiros replied:
Yes, you can't analyze the delta object in a separate thread. The delta object has to be analyzed in the listener thread (the thread that triggers the resource listener). But what I was saying is that the listener analysis should consist only of marking locations as having out-of-date structure/index. You don't actually calculate the new structure/index in the listener thread. Rather, when the index is marked as out of date for a location, that triggers a recalculation in a background thread. Again the DocumentReconcileManager would be a good example of a such a design. But it is generally more or less what you said, scheduling one job per file/location. (with a few more minor details)

bruno-medeiros replied:
Note, don't use an Eclipse Job, as there is no point to that. Eclipse Jobs are good for user operations/tasks, that the user itself can see in the Progress view, and optionally cancel, but this is an internal task. Use a Java Executor instead and submit a task there. See `melnorme.lang.ide.core.engine.AbstractAgentManager` for example...

Woyten replied:
This should be solved by now. The delta event is processed synchronously, i.e. the index updates are enqueued using the `ConcurrentlyDerivedData` mechanism. The `ConcurrentlyDerivedData` instances are remembered per file, s.t. a second thread working on the same file will request the cancellation of the first one.
The remembered `ConcurrentlyDerivedData` are not removed yet. They do not harm but consume memory. They could be remove whenever a thread has finished but that would require synchronizing which could slow down the indexing performance.

Woyten replied:
By the way, `RustStructureUpdateTask` and `RustIndexUpdateTask` do more or less the same thing. I would like to turn them into one class but that would require touching additional classes, e.g. `StructureInfo`. Any problems with that?

Woyten replied:
I managed to unify the task classes. Had to touch the surrounding infrastructure a bit but I think the code is a bit cleaner now as it no longer requires the dependency of the nested non-static `StructureInfo` class.

Woyten commented:
@bruno-medeiros Can this be merged now? To my best knowledge, I addressed every point you mentioned in the review. In addition, I added two basic tests but, in order to test the concurrent updates due to resource changes, we would need some additional testing framework like `Mockito`. I also adapted the release notes and enhanced the _Navigate_ menu.

bruno-medeiros commented:
Ok, first of all check my other comment in the other PR: https://github.com/RustDT/RustDT/pull/139#issuecomment-258490998 , especially the part regarding RLS.

With that in mind, there's a big issue have with this functionality: it will only index files for projects in the Eclipse workspace. But it will not index symbols in files from dependency crates (nor the standard lib). This in my view is a big limitation. 

This could be addressed by invoking `cargo metadata` which will list all the (transitive) dependencies of a crate, and their location in the filesystem (and download them if necessary). This would make it easy to then search for .rs files and index all symbols.

However... this adds more complexity to the implementation of this feature, and yet, the core functionality is likely to be made obsolete by RLS itself providing the symbols search functionality. There is already support for this in the LSP protocol ( https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#workspace-symbols-request ) - although RLS doesn't support that functionality yet.

However, if this is to be implemented fully (symbol search across all dependency crates), one can make the case that it would be much better to implement it in RLS directly. Anything else will just be a temporary measure, and Eclipse-specific, whilst implementing in RLS itself would benefit a much larger audience of people using other IDEs.

I don't know if that is something that interests you or not, but for me the big priority is to get RLS into a more usable shape (less alpha), and then to integrate RLS/LSP support into Eclipse/RustDT. Spending time and effort in functionality that could be made obsolete by RLS integration seems like a waste at this point.

Woyten commented:
You're right. Not every file is indexed because some of the source files are copied to the cargo cache folder. Strange... I did not observe this behavior before. Probably this has changed after some rust version update... Then, indeed, it's getting difficult as the relevant set of sources depends on the `cargo metadata` result of all `cargo.toml`s. As a workaround the `cargo` cache folder could be added as a scan folder but then the index will probably see too many versions and unused source files.

Woyten committed:
Implement open type feature

Woyten commented:
I learned that it's a common practice to update PRs via force pushes in order to keep them small.

My PRs are now significantly smaller, just in case you are interested,

bruno-medeiros commented:
@Woyten I appreciate the effort, but I'm also closing this PR: as explained before, I don't think it would be good to add any new semantic/indexing functionality that is not tied to LSP. In other words this should all be done by the RLS https://github.com/jonathandturner/rls

At most, you the UI part of this PR (the dialog, etc.) could be re-used, but the engine itself would have to be LSP based.
